### PR TITLE
Bug 1344358 - Make perf regression detection work better w/ retriggers

### PR DESCRIPTION
**Pull request by wlach**
Retriggering performance jobs to generate better performance numbers
is becoming increasingly common, but our old regression detection
code didn't really account for that case -- in particular, it
had a bug that would allow for a regression to be generated *inside*
the data for a particular revision (if there were enough).

This commit fixes that bug and also changes the behaviour of the
linear weighting function so that retriggers on a particular revision
will be weighted the same (rather than declining in importance)

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2284)

### CODE REVIEW

File: treeherder/perfalert/perfalert/__init__.py [hunk 1]
```diff
@@ -55,104 +72,107 @@ def calc_t(w1, w2, weight_fn=None):
...
-        kw = [d.value for d in data[i:i+k]]
 
-        # Reverse the backward data so that the current point is at the
-        # start of the window.
-        jw.reverse()
+        # keep on getting previous data until we've either got at least 12
+        # data points *or* we've hit the maximum back window
+        jw = []
+        di.amount_prev_data = 0
+        prev_indice = (i - 1)
```
Comment (jmaher) [inline treeherder/perfalert/perfalert/__init__.py]: I am not used to seeing a calculation wrapped in (), is this a convention inside of treeherder?

File: treeherder/perfalert/perfalert/__init__.py [hunk 2]
```diff
@@ -55,104 +72,107 @@ def calc_t(w1, w2, weight_fn=None):
...
+
+        # data values associated with this revision
+        self.values = copy.copy(values)
 
         # t-test score
         self.t = 0
+
         # Whether a perf regression is found
-        self.state = state
+        self.state = 'good'
```
Comment (jmaher) [inline treeherder/perfalert/perfalert/__init__.py]: I don't understand the hardcoded state of good here.

File: treeherder/perf/alerts.py [hunk 3]
```diff
@@ -65,19 +70,16 @@ def generate_new_alerts_in_series(signature):
...
     with transaction.atomic():
         for (prev, cur) in zip(analyzed_series, analyzed_series[1:]):
-            # we can have the same testrun id in a sequence if there are
-            # retriggers, so only set the prev_testrun_id if that isn't
-            # the case
-            if prev.testrun_id != cur.testrun_id:
-                prev_testrun_id = prev.testrun_id
-
-            if cur.state == 'regression' and prev_testrun_id:
+            print (prev, cur)
```
Comment (jmaher) [inline treeherder/perf/alerts.py]: we are printing stuff here, but where does it go?  Is this for local debugging?

File: treeherder/perfalert/perfalert/__init__.py (hunk 1 above)
Comment (wlach) [inline treeherder/perfalert/perfalert/__init__.py]: No, just an arbitrary decision from me. Probably not necessary, I'll take out the parans.

File: treeherder/perfalert/perfalert/__init__.py (hunk 2 above)
Comment (wlach) [inline treeherder/perfalert/perfalert/__init__.py]: Silly legacy from old perf regression code. I'll just change this to a boolean called `change_detected`, which should be much easier to understand.

File: treeherder/perf/alerts.py (hunk 3 above)
Comment (wlach) [inline treeherder/perf/alerts.py]: Yeah, this was left in by mistake. Will fix.

**Pull request merged**